### PR TITLE
Abstract Source Open Workflows

### PR DESCRIPTION
**Pull request by Trenly**
- [x] I have signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs).
- [ ] This pull request is related to an issue.

It seemed to me that the way sources are opened is very repetitive across commands, and especially within subcommands. If this is something that isn't wanted, please close it, just thought I would make the PR and see where it went.

-----

 ###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/3224)

### CODE REVIEW

File: src/AppInstallerCLICore/Commands/InstallCommand.cpp [hunk 1]
```diff
@@ -117,7 +117,7 @@ namespace AppInstaller::CLI
         {
             context <<
                 Workflow::ReportExecutionStage(ExecutionStage::Discovery) <<
-                Workflow::OpenSource();
+                Workflow::OpenSourceFor(OperationType::Install);
 
             if (!context.Args.Contains(Execution::Args::Type::Force))
```
[the remark this answers is not in the document]
Comment (Trenly) [reply src/AppInstallerCLICore/Commands/InstallCommand.cpp]: Ah, I didn't see that down there. Should be resolved

File: src/AppInstallerCLICore/Workflows/WorkflowBase.cpp [hunk 2]
```diff
@@ -444,6 +444,42 @@ namespace AppInstaller::CLI::Workflow
...
+        case OperationType::List:
+        case OperationType::Uninstall:
+        case OperationType::Upgrade:
+            context <<
+                Workflow::OpenSource() <<
+                Workflow::OpenCompositeSource(Workflow::DetermineInstalledSource(context));
+            break;
+        case OperationType::Pin:
+        case OperationType::Export:
+        case OperationType::Completion:
```
Comment (florelis) [inline src/AppInstallerCLICore/Workflows/WorkflowBase.cpp:473]: I'm not sure about `Completion`. There are other cases like search where we don't use the installed, and I see you had to use Search type there

File: src/AppInstallerCLICore/Commands/PinCommand.cpp [hunk 3]
```diff
@@ -119,14 +117,12 @@ namespace AppInstaller::CLI
             // When we are given an ID, just pin that available package without checking for installed.
             // This helps when there are matching issues, for example due to multiple side-by-side installs.
             context <<
-                Workflow::OpenSource();
+                Workflow::OpenSourceFor(OperationType::Search);
```
Comment (florelis) [inline src/AppInstallerCLICore/Commands/PinCommand.cpp:120]: Also not a fan of having to use ::Search in the Pin command

File: src/AppInstallerCLICore/Commands/PinCommand.cpp [hunk 4]
```diff
@@ -168,9 +164,7 @@ namespace AppInstaller::CLI
 
     void PinRemoveCommand::Complete(Execution::Context& context, Args::Type valueType) const
     {
-        context <<
-            Workflow::OpenSource() <<
-            Workflow::OpenCompositeSource(Repository::PredefinedSource::Installed);
+        context << Workflow::OpenSourceFor(OperationType::Pin);
```
Comment (florelis) [inline src/AppInstallerCLICore/Commands/PinCommand.cpp:167]: I forgot to change this. Maybe we should also conditionally open only the available source here

File: src/AppInstallerCLICore/Workflows/WorkflowBase.cpp (hunk 2 above)
Comment (Trenly) [reply src/AppInstallerCLICore/Workflows/WorkflowBase.cpp:473]: Hmm. I can understand that point. One thing that I don't understand is the difference between the `PredefinedSource::Installed` and `DetermineInstalledSource`. The cases where completion uses the Predefined source are for the commands which use `DetermineInstalledSource`. 

Perhaps what would make the most sense is to have completion throw the invalid state and keep the current behavior of explicit management of sources for completion

File: src/AppInstallerCLICore/Commands/PinCommand.cpp (hunk 3 above)
Comment (Trenly) [reply src/AppInstallerCLICore/Commands/PinCommand.cpp:120]: Understandable; Will adjust

File: src/AppInstallerCLICore/Workflows/WorkflowBase.cpp (hunk 2 above)
Comment (yao-msft) [reply src/AppInstallerCLICore/Workflows/WorkflowBase.cpp:473]: DetermineInstalledSource will filter the installed source to include only user scope | machine scope | default (previous behavior where for win32, user and machine scopes are included, for msix, only apps registered to user are included)

**Pull request closed without merging**